### DISCUSSION TITLE
video capturing - high CPU usage

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/meta-tegra#704] [General]
Hello,
We are using the l4t, 32.5.0 branch, for capturing video from a 13mp camera (4192*3120) to v4l2.
We get a high CPU usage (100% of a single core) for the capturing as well as video is not well streamed for ~10 minutes.
After 10 minutes the CPU usage goes down to ~30% and video goes pretty well.

I tried looking at the gstreamer logs but didn't see anything suspicious, looking at both when it lags and when it starts playing right, logs looks the same.

I am trying to understand what is the flow I should be checking (v4l2, gstreamer, omx), any specific hw acceleration I should be looking for etc.

Anyone aware of such behavior?
Where should I start? Maybe this should go to the gstreamer folks?

Thanks

**Comment by tirankas** [comment]
Adding:
Running: gst-inspect-1.0 |grep omx
I get nothing.

Running the same on the Nvidia Ubuntu image I get several omx.

What am I missing?
hwcodecs are part of my image recipe, gstreamer1.0-omx-tegra is being picked up in my build and the crresponding lib is part of my image.

Thanks

**Reply by madisongh** [reply]
You should start by finding out why it's been blacklisted:
```
$ rm -rf ~/.cache/gstreamer-1.0
$ GST_DEBUG_NO_COLOR=1 GST_DEBUG=4 gst-inspect-1.0 -b &> /tmp/gst.log
```
Then inspect the log file to see what errors occurred when it tried to load the omx plugin.

**Comment by tirankas** [comment]
Hi Matt,
I found a fix you did long ago (https://github.com/OE4T/meta-tegra/commit/64600fdaec296010313726aaf150d84efcc732e8#diff-f60f85ee8e4ca2b4fe637b18a7676ce6afee64e36047656b62f0f034be661598) for the exact same issue.
I see this patch is no longer taken and wonder how did you conclude the issue in the first place so that I can check myself on the latest branch.

Thanks

**Reply by madisongh** [reply]
That was a very long time ago, in a pre-release version of the BSP. I think NVIDIA fixed the build problem that caused me to add that patch as a workaround.

**Comment by tirankas** [comment]
Progressing a bit:

gst-inspect-1.0 libgstomx.so
Plugin Details:
  Name                     libgstomx.so
  Description              Plugin for blacklisted file
  Filename                 /usr/lib/gstreamer-1.0/libgstomx.so
  Version                  0.0.0
  License                  BLACKLIST
  Source module            BLACKLIST
  Binary package           BLACKLIST
  Origin URL               BLACKLIST


  0 features:

Where is this blacklist coming from?

Also this:
sudo gst-inspect-1.0 /usr/lib/gstreamer-1.0/libgstomx.so
Plugin Details:
  Name                     omx
  Description              GStreamer OpenMAX Plug-ins
  Filename                 /usr/lib/gstreamer-1.0/libgstomx.so
  Version                  1.0.0
  License                  LGPL
  Source module            gst-omx
  Source release date      2013-03-22
  Binary package           GStreamer OpenMAX Plug-ins source release
  Origin URL               Unknown package origin

  omxmpeg4videodec: OpenMAX MPEG4 Video Decoder
  omxh264dec: OpenMAX H.264 Video Decoder
  omxh265dec: OpenMAX H.265 Video Decoder
  omxvp8dec: OpenMAX VP8 Video Decoder
  omxvp9dec: OpenMAX VP9 Video Decoder
  omxmpeg2videodec: OpenMAX MPEG2 Video Decoder
  omxwmvdec: OpenMAX WMV Video Decoder
  omxh264enc: OpenMAX H.264 Video Encoder
  omxh265enc: OpenMAX H.265 Video Encoder
  omxvp8enc: OpenMAX VP8 Video Encoder
  omxvp9enc: OpenMAX VP9 Video Encoder
  nvoverlaysink: OpenMax Video Sink

  12 features:
  +-- 12 elements

**Reply by madisongh** [reply]
See my comment above.  gstreamer will blacklist any plugin that it can't load and initialize.  Looking at the error messages that get produced during that process should point you in the right direction.

**Comment by madisongh** [comment]
I'd also recommend looking at the [Accelerated GStreamer](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/accelerated_gstreamer.html#) chapter of the L4T documentation for more information about the tegra-specific plugins.

**Comment by tirankas** [comment]
HI Matt,
Apparently there was no issue with the omx plugin.
It doesn't list omx when running the command without sudo.
Running the gst inspect command with a sudo shows omx.

Back to square 1, still need to figure out why capturing from a camera (v4l2) consumes 90% cpu...

Thanks

**Reply by dwalkes** [reply]
@tirankas you might want to try cross-posting on the NVIDIA developer forum if you haven't already.  I typically reproduce with the stock L4T release and no customizations first.
